Don't duplicate snippets in "Snippets" menu

chosak opened the pull request:
This change alters the behavior of the sidebar "Snippets" menu such that the snippet index view doesn't include snippet models that have been configured with their own menu item.

Currently, the "Snippets" menu doesn't appear if all snippet models have their own menu items; however, if some models do and some don't, the snippets index view lists all of them, causing some duplication in the UI which could be confusing for editors.

This commit changes this behavior so that "Snippets" is only used to edit those snippet models that aren't editable elsewhere.

A new setting (`WAGTAILSNIPPETS_MENU_SHOW_ALL`) has been added to always show the snippets menu and all snippets models. This setting will allow for continuation of the current behavior for those users that desire it.

Fixes #11340.

(Note: we could also be more aggressive here and not add a new setting, and just change the default behavior as described -- personally I think this makes more sense than always showing everything in "Snippets". I included the new setting here just as a way to maintain backwards compatibility but would be happy to alter the PR to remove it as well.)

## Screenshots of documentation changes:

### Customising admin views for snippets

![image](https://github.com/wagtail/wagtail/assets/654645/2eb0a9f1-b4ac-477a-9872-9fd2d3e41f8f)

### Settings

![image](https://github.com/wagtail/wagtail/assets/654645/7fc16dcf-ed3d-4411-9aeb-42a380941bb9)

### Wagtail 6.0 release notes - IN DEVELOPMENT

![image](https://github.com/wagtail/wagtail/assets/654645/760d7414-5aeb-4ddb-a90d-a6f71aed7340)

lb- commented:
I think this is looking good. I wonder if we should maybe reverse the default and then add upgrade considerations. Depending on timing for 6.0 though.

laymonage commented:
Drive-by comment: I think the breadcrumbs will still refer to snippets.

chosak commented:
@lb- by reversing the default do you mean keeping the current behavior and gradually switching over to the new behavior? FWIW I've rebased PR this to main.

@laymonage thank you for the valuable drive-by comment. If with this PR there would no longer be a snippets index view (because all snippets are editable elsewhere in the admin), then you're right that it wouldn't make sense to include that in the breadcrumbs.

Currently, if all snippets have their own menu items (as in the default bakerydemo config), "Snippets" doesn't exist in the menu but still exists in individual snippet breadcrumbs, and users can still access /admin/snippets/ even though it isn't accessible from the menu.

I could push a change that removes "Snippets" from the breadcrumbs if that page doesn't need to exist, but /admin/snippets/ would return 403 in that case, even though /admin/snippets/something/ might not.

I could preserve the current behavior where /admin/snippets/ still functions but just isn't linked from anywhere, but that doesn't feel exactly right either - then again that's how it works now in bakerydemo.. what do you think?

laymonage reviewed (commented):
Apologies for missing this @chosak. I don't think the `/admin/snippets/` URL needs to be made inaccessible ([cool URIs don't change](https://www.w3.org/Provider/Style/URI)).

I'm just saying that if a model were hidden on the snippets model listing, then the breadcrumbs items of the model's views shouldn't link to the snippets model listing, as you technically couldn't go to the model's views from there.

I think it's fine to have the `/admin/snippets/` URL show an empty list when all models have their own menu items (and the new setting is set to `False`). Users won't be able to get to this URL normally, but if they manually enter the URL then it will just render an empty list, similar to how index views work when there are no items.

chosak commented:
@laymonage I've modified the snippet list view breadcrumbs to exclude "Snippets" if a snippet has its own menu item and thus is not listed there.

> I think it's fine to have the /admin/snippets/ URL show an empty list when all models have their own menu items (and the new setting is set to False). Users won't be able to get to this URL normally, but if they manually enter the URL then it will just render an empty list, similar to how index views work when there are no items.

I'm a little unsure about this. I get the "cool URIs don't change" logic. But if a user can access /admin/snippets/something/, and, with this PR, that page no longer has breadcrumbs pointing to /admin/snippets/, and the only way a user can get to that URL is by manually entering it, and that page no longer lists any of the snippets because they all have their own menu items, our options are:

1. /admin/snippets/ redirects users to /admin/ with the "Sorry, you do not have permission to access this area." message that we show normally if there are no snippets to edit at all, or
2. /admin/snippets/ shows an empty list with an unexpected "There are no results." message.

Neither of these seem great to me but the empty list especially seems like something has gone wrong. At least in the "do not have permission" case we're being consistent about /admin/snippets/ not actually being a valid view. What if we changed that error message to something more accurate, or removed the error and just had a plain redirect to home without any message at all?

funkybob commented:
Will this also omit Snippets that have their viewsets in a `SnippetViewSetGroup` ?

gasman commented:
@funkybob Yes - this is reusing the existing `menu_item_is_registered` flag which is currently used to determine whether to show the top-level Snippets menu item, and takes SnippetViewSetGroup into account:

https://github.com/wagtail/wagtail/blob/714e5c0f62d6c024450689073faac1783a27386b/wagtail/snippets/views/snippets.py#L1182-L1185

gasman commented:
Re accessibility of the `/admin/snippets/` view - not sure if this was already the case when this PR was made or whether it's new as of #12013, but there's an existing check in ModelIndexView to respond with PermissionDenied if the list of snippet types is empty, so this naturally extends to the new case introduced by this PR, where the list is empty because all snippet types are editable through their own menu items.

I don't see any problem with that (or for that matter returning an empty list) - with the link gone from the breadcrumb, there's now essentially nothing left in the user interface to suggest that the models they have access to should be regarded as "snippets" (other than the URL path - and anyone who's enough of a power user to hack the URL directly should be immune to those sorts of surprises :-) ), and so a user should have no particular expectations about what ought to be on a list of snippet types.

(incidentally, #12013 makes this change https://github.com/wagtail/wagtail/pull/11401#discussion_r1568505227 redundant - I'm about to push a rebase of this PR, and leaving that commit out.)

gasman commented:
Rebased in #13031
